### PR TITLE
Implement support for Id field type.

### PR DESCRIPTION
**Pull request by BearLemma**
Adds support for `Id` type which can be used to model relationships.

### CODE REVIEW

File: api/src/datastore.ts [hunk 1]
```diff
@@ -933,6 +934,19 @@ export class ChiselEntity {
         return undefined;
     }
 
+    static async byId<T extends ChiselEntity>(
+        this: { new (): T },
+        entity_id: Id<T>,
+    ): Promise<T | undefined> {
```
Comment (honzasp) [inline api/src/datastore.ts:940]: I suggest that we provide two variants of this method, first that returns `T` or throws, and second that returns `T | undefined`. In my opinion, the first variant is going to be much more common, because there are going to be few cases where you look for an entity by its id and expect that it might not exist (for the same reason that there are few cases where you have a pointer and expect that the pointed-to object might not exist).

So I propose that the short `byId()` return `T`, and a longer variant, `byIdOptional()` or `findById()` or `getById()` or whatever,  return `T | undefined`.

File: cli/src/ts.rs [hunk 2]
```diff
@@ -149,6 +152,30 @@ fn map_type(handler: &Handler, x: &TsType) -> Result<TypeEnum> {
     }
 }
 
+fn map_entity_id(handler: &Handler, tr: &TsTypeRef) -> Result<TypeEnum> {
+    let type_params = &tr
+        .type_params
+        .as_ref()
+        .context("Id field declared without Entity parameter")?
```
Comment (honzasp) [inline cli/src/ts.rs:159]: I would reword this message to more accurately reflect the error, something like "Type `Id` requires type arguments".

File: cli/src/ts.rs [hunk 3]
```diff
@@ -149,6 +152,30 @@ fn map_type(handler: &Handler, x: &TsType) -> Result<TypeEnum> {
...
+        .params;
+    ensure!(
+        type_params.len() == 1,
+        "Unexpected number of Id generic parameters"
```
Comment (honzasp) [inline cli/src/ts.rs:163]: Similar to above, I would reword it something like "Type `Id` requires exactly one type argument", to be similar to error messages produced by `tsc`.

File: cli/src/ts.rs [hunk 4]
```diff
@@ -149,6 +152,30 @@ fn map_type(handler: &Handler, x: &TsType) -> Result<TypeEnum> {
...
+    if let TsType::TsTypeRef(tr) = &**param {
+        if let TsEntityName::Ident(entity_id) = &tr.type_name {
+            let entity_name = ident_to_string(entity_id);
+            return Ok(TypeEnum::EntityId(entity_name));
+        }
+    }
+    Err(swc_err(
+        handler,
+        tr,
+        "failed to parse entity name out of Id field",
```
Comment (honzasp) [inline cli/src/ts.rs:175]: What about "The type argument of type `Id` must refer to an entity"?

File: cli/tests/integration_tests/rust_tests/model_ids.rs [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+            const trial = await Book.findOne({title: "The Trial"});
+            const author = await Author.byId(trial!.author);
+            return author!.name;
+        }"#,
+    );
+
+    c.chisel.apply_ok().await;
+    c.chisel.post("/dev/store").send().await.assert_ok().text();
+    c.chisel
+        .get("/dev/get_joke_author")
```
Comment (honzasp) [inline cli/tests/integration_tests/rust_tests/model_ids.rs:50]: Probably `get_trial_author`? :)

File: cli/tests/integration_tests/rust_tests/model_ids.rs [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+        "models/models.ts",
+        r#"
+        import { ChiselEntity, Id } from "@chiselstrike/api";
+
+        export class Book extends ChiselEntity {
+            title: string;
+            author: Id<Author>;
+        }"#,
+    );
+    c.chisel.apply_err().await;
```
Comment (honzasp) [inline cli/tests/integration_tests/rust_tests/model_ids.rs:119]: I think we should assert that it failed with an expected error, rather than for some unexpected reason.

File: server/src/ops/type_system.rs [hunk 7]
```diff
@@ -57,6 +57,10 @@ enum SimpleTypeId {
         #[serde(rename = "entityName")]
         entity_name: String,
     },
+    EntityId {
+        #[serde(rename = "entityId")]
```
Comment (honzasp) [inline server/src/ops/type_system.rs:61]: Did you mean `rename = "entityName"`?

File: server/src/types/mod.rs [hunk 8]
```diff
@@ -20,6 +20,7 @@ pub enum Type {
     Boolean,
     JsDate,
     Entity(Entity),
+    EntityId { entity_name: String },
```
Comment (honzasp) [inline server/src/types/mod.rs:23]: It's a bit weird when one enum mixes tuple variants `Entity(Entity)` and struct variants `EntityId { entity_name: String }`, do you think that `EntityId(String)` would be unclear?

**Review by honzasp — COMMENTED** [review]
Looks ok, I have just some minor comments, nothing really important.

**Review by MarinPostma — COMMENTED** [review]
lgtm

**Comment by MarinPostma** [comment]
I'm thinking that you could add a test that ensure that an entity `A` cannot be queried with an `Id<B>`?

**Comment by MarinPostma** [comment]
right, but that defeats the purpose of typed ids a bit

File: cli/tests/integration_tests/rust_tests/model_ids.rs [hunk 9]
```diff
@@ -0,0 +1,178 @@
...
+
+#[self::test(modules = Deno)]
+pub async fn basic(c: TestContext) {
+    write_models(&c.chisel);
+    c.chisel.write(
+        "routes/store.ts",
+        r#"
+        import { Book, Author } from '../models/models.ts';
+        export default async function () {
+            const kafka = await Author.create({
```
Comment (glommer) [inline cli/tests/integration_tests/rust_tests/model_ids.rs:27]: are we going to support RedPanda as well ?

**Review by honzasp — APPROVED** [review]
